sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -161,7 +182,20 @@ void Scheduler::removeMemoryObject(detail::SYCLMemObjI *MemObj) {
 
 EventImplPtr Scheduler::addHostAccessor(Requirement *Req,
                                         const bool destructor) {
-  std::lock_guard<std::shared_timed_mutex> Lock(MGraphLock);
+  std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+  // Avoiding deadlock situation for MSVC. std::shared_timed_mutex specification
+  // does not specify a priority for shared and exclusive accesses. It will be a
+  // deadlock in MSVC's std::shared_timed_mutex implementation, if exclusive
+  // access occurs after shared access.
+  // TODO: after switching to C++17, change std::shared_timed_mutex to
+  // std::shared_mutex and use std::lock_guard here both for Windows and Linux.
+#ifdef _WIN32
+  while (!Lock.owns_lock()) {
+    Lock.try_lock();
+  }
+#else
+  Lock.lock();
+#endif // _WIN32
```

Comment:
Can we have a function/macro for this to avoid copy-paste of this code?
Sure thing, the comment should stay near the function/macro.

Reply:
Fixed